onadata/apps/django_digest_backends/cache.py

```diff
@@ -0,0 +1,54 @@
+from django.core.cache import caches
+from django_digest.utils import get_setting
+
+
+DIGEST_NONCE_CACHE_NAME = get_setting('DIGEST_NONCE_CACHE_NAME', 'default')
+NONCE_TIMEOUT = get_setting('DIGEST_NONCE_TIMEOUT_IN_SECONDS', 5 * 60)
+NONCE_NO_COUNT = ''  # Needs to be something other than None to determine not set vs set to null
+
+
+class RedisCacheNonceStorage():
+    def _get_cache(self):
+        # Dynamic fetching of cache is necessary to work with override_settings
+        return caches[DIGEST_NONCE_CACHE_NAME]
+
+    def _generate_cache_key(self, user, nonce):
+        return f'user_nonce_{user}_{nonce}'
+
+    def update_existing_nonce(self, user, nonce, nonce_count):
+        """
+        Check and update nonce record. If no record exists or has an invalid count,
+        return False. Create a lock to prevent a concurrent replay attack where
+        two requests are send immediately and either may finish first.
+        """
+        cache = self._get_cache()
+        cache_key = self._generate_cache_key(user, nonce)
+
+        if nonce_count == None:  # No need to lock
+            existing = cache.get(cache_key)
+            if existing is None:
+                return False
+            cache.set(cache_key, NONCE_NO_COUNT, NONCE_TIMEOUT)
+        else:
+            with cache.lock(
+                f'user_nonce_lock_{user}_{nonce}',
+                timeout=NONCE_TIMEOUT,
+                blocking_timeout=30
```

Comment:
just curious how this works. is this right?
1. process a takes the lock and somehow dies without releasing it
2. process b tries to get the lock and fails with a `LockError` after 30 seconds
3. process c tries to get the lock and fails with a `LockError` after 30 seconds
4. etc., until the TTL of the lock key (set to `NONCE_TIMEOUT`) expires?

if so, could we catch `LockError` and delete the nonce outright to get back to a better state and minimize the number of worker processes stuck waiting for the lock?

Reply:
Implemented. Give this a test please. I used KoboCollect and deleted the nonces from redis. From earlier testing, it seems like it doesn't matter to enketo or collect if we delete the nonces.